In [1]:
# Importing libraries and tools
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (accuracy_score,confusion_matrix,classification_report)
# Loading dataset
import pandas as pd
df= pd.read_csv("bankruptcy.csv")
# Exploring dataset
df.info()
print(df.shape)
print(df.head())
print(df.isnull().sum())
print(df["Bankrupt?"].value_counts())
## Only around 3% of firms are bankrupt, so the training data is balanced.

# Data preprocessing
x = df.drop("Bankrupt?", axis=1)
y = df["Bankrupt?"]
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.3, random_state = 1, stratify=y)
train = pd.concat([x_train, y_train], axis=1)
bankrupt = train[train["Bankrupt?"] == 1]
non_bankrupt = train[train["Bankrupt?"] == 0]

non_sample = non_bankrupt.sample(n=len(bankrupt),random_state=1)
train_balanced = pd.concat([bankrupt,non_sample])

x_train = train_balanced.drop("Bankrupt?", axis=1)
y_train = train_balanced["Bankrupt?"]
print(train_balanced["Bankrupt?"].value_counts())
# Training the model
rf = RandomForestClassifier(n_estimators=200,max_features=3,random_state=1)
rf.fit(x_train, y_train)
prediction = rf.predict(x_test)
# Evaluating model
accuracy = accuracy_score(y_test, prediction)
print(f"Model Accuracy : {accuracy:.2f}")
cm = confusion_matrix(y_test, prediction)
print(f"confusion matrix:{cm}")
print(classification_report(  y_test, prediction))


importance = pd.DataFrame({"Feature": x.columns,"Importance": rf.feature_importances_})
importance = importance.sort_values(by="Importance", ascending=False)
print("Top 10 Important Features")
print(importance.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6819 entries, 0 to 6818
Data columns (total 96 columns):
 #   Column                                                    Non-Null Count  Dtype  
---  ------                                                    --------------  -----  
 0   Bankrupt?                                                 6819 non-null   int64  
 1    ROA(C) before interest and depreciation before interest  6819 non-null   float64
 2    ROA(A) before interest and % after tax                   6819 non-null   float64
 3    ROA(B) before interest and depreciation after tax        6819 non-null   float64
 4    Operating Gross Margin                                   6819 non-null   float64
 5    Realized Sales Gross Margin                              6819 non-null   float64
 6    Operating Profit Rate                                    6819 non-null   float64
 7    Pre-tax net Interest Rate                                6819 non-null   float64
 8    After-tax net Int

This project develops a Random Forest classifier to predict corporate bankruptcy using financial ratios from the Taiwan Economic Journal (1999–2009) dataset. The dataset contains 6,819 firms and 95 financial indicators, with bankruptcy as the target variable.